mlflow/utils/autologging/logging_and_warnings.py

```diff
@@ -0,0 +1,287 @@
+import os
+import warnings
+from contextlib import contextmanager
+from pathlib import Path
+from threading import RLock, get_ident as get_current_thread_id
```

Comment:
Was wondering about the performance of `get_current_thread_id` since we call it every time a warning is logged & @harupy previously found that other Python introspection methods (e.g. getting the current call stack) were rather slow e.g. took ~1ms - seems like it's super fast:

```
>>> def benchmark(n):
...   s = time.time()
...   for i in range(n):
...     get_current_thread_id()
...   print("Took %s sec" % (time.time() - s))
... 
>>> benchmark(1000)
Took 0.00011706352233886719 sec
>>> benchmark(10000)
Took 0.0009729862213134766 sec
>>> benchmark(100000)
Took 0.008674144744873047 sec
```

In general seems reasonable that logging a warning (writing to some non-memory output stream) would be quite a bit slower than the in-memory checks we do to determine whether to log the warning